### PR TITLE
tests/malloc_thread_safety: improve test

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- fix false test failure on picolibc
    - For picolibc, the result of mallinfo().uordblks doesn't go back
      to zero after the first allocation. By adding `free(malloc(4))`
      before recording the pre-test mallinfo() value, this issues is
      worked arround
- increase test coverage to also cover realloc()
- ease testing with picolibc by allowing `make PICOLIBC=1 flash test`

### Testing procedure

Test application should still pass. After applzing

```diff
diff --git a/cpu/cortexm_common/Makefile.dep b/cpu/cortexm_common/Makefile.dep
index 856befb6c1..309d8c9a9a 100644
--- a/cpu/cortexm_common/Makefile.dep
+++ b/cpu/cortexm_common/Makefile.dep
@@ -25,4 +25,4 @@ ifeq (1, $(DEVELHELP))
 endif
 
 # Make calls to malloc and friends thread-safe
-USEMODULE += malloc_thread_safe
+#USEMODULE += malloc_thread_safe
```

the test should crash.

### Issues/PRs references

Came up in https://github.com/RIOT-OS/RIOT/pull/15999

### CODE REVIEW

**Comment by maribu** [comment]
Test results

##### This PR, with commenting out module `malloc_thread_safe` (expecting failure)

```
$ make BOARD=nucleo-f767zi flash test -C tests/malloc_thread_safety/

[...]

Test Application for multithreaded use of malloc()
==================================================

This test will run duelling threads allocating and freeing memory.
The running thread is interrupted every millisecond and the other
threads gets scheduled. Eventually, this should yield to memory
corruption unless proper guards are in place preventing them. After
ca. two seconds without crash, the test is considered as passing.

Testing: malloc()/free()

Context before hardfault:
   r0: 0x20000224
   r1: 0x00000020
   r2: 0x00000000
   r3: 0x20000220
  r12: 0x40000030
   lr: 0x080002a9
   pc: 0x08001c88
  psr: 0x81000000

FSR/FAR:
 CFSR: 0x00000400
 HFSR: 0x40000000
 DFSR: 0x0000000b
 AFSR: 0x00000000
Misc
EXC_RET: 0xfffffffd
Active thread: 2 "t1"
Attempting to reconstruct state for debugging...
In GDB:
  set $pc=0x8001c88
  frame 0
  bt
```

```
$ make BOARD=nucleo-f767zi PICOLIBC=1 flash test -C tests/malloc_thread_safety/

[...]

main(): This is RIOT! (Version: 2021.04-devel-1143-g51908-tests/malloc_thread_safety)
Test Application for multithreaded use of malloc()
==================================================

This test will run duelling threads allocating and freeing memory.
The running thread is interrupted every millisecond and the other
threads gets scheduled. Eventually, this should yield to memory
corruption unless proper guards are in place preventing them. After
ca. two seconds without crash, the test is considered as passing.

Testing: malloc()/free()
Timeout in expect script at "child.expect("TEST PASSED")" (tests/malloc_thread_safety/tests/01-run.py:16)
```

##### This PR, as is (expecting pass)

```
$ make BOARD=nucleo-f767zi flash test -C tests/malloc_thread_safety/

[...]

main(): This is RIOT! (Version: 2021.04-devel-1143-g51908-tests/malloc_thread_safety)
Test Application for multithreaded use of malloc()
==================================================

This test will run duelling threads allocating and freeing memory.
The running thread is interrupted every millisecond and the other
threads gets scheduled. Eventually, this should yield to memory
corruption unless proper guards are in place preventing them. After
ca. two seconds without crash, the test is considered as passing.

Testing: malloc()/free()
Testing: realloc()/free()
TEST PASSED

make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/malloc_thread_safety'
```

```
$ make BOARD=nucleo-f767zi PICOLIBC=1 flash test -C tests/malloc_thread_safety/

[...]

main(): This is RIOT! (Version: 2021.04-devel-1143-g51908-tests/malloc_thread_safety)
Test Application for multithreaded use of malloc()
==================================================

This test will run duelling threads allocating and freeing memory.
The running thread is interrupted every millisecond and the other
threads gets scheduled. Eventually, this should yield to memory
corruption unless proper guards are in place preventing them. After
ca. two seconds without crash, the test is considered as passing.

Testing: malloc()/free()
Testing: realloc()/free()
TEST PASSED

make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/malloc_thread_safety'
```

**Comment by fjmolinas** [comment]
@maribu any reason you did not mark this one as ready-for build? I can run the tests on this PR, and the changes make sense.

**Comment by maribu** [comment]
> any reason you did not mark this one as ready-for build?

Just forgot to do :-)

File: tests/malloc_thread_safety/Makefile [hunk 1]
```diff
@@ -1,9 +1,15 @@
 include ../Makefile.tests_common
 
+PICOLIBC ?= 0
+ifneq (0,$(PICOLIBC))
+  USEMODULE += picolibc
+  FEATURES_REQUIRED += picolibc
```
Comment (fjmolinas) [inline tests/malloc_thread_safety/Makefile:6]: Isn't only `FEATURES_REQUIRED += picolibc` needed?

File: tests/malloc_thread_safety/Makefile [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
+endif
 USEMODULE += xtimer
 
 include $(RIOTBASE)/Makefile.include
 
+
```
Comment (fjmolinas) [inline tests/malloc_thread_safety/Makefile:12]: unrelated

File: tests/malloc_thread_safety/main.c [hunk 3]
```diff
@@ -57,6 +57,24 @@ void * t1_t2_func(void *arg)
     return NULL;
 }
 
+void * t1_t2_realloc_func(void *arg)
+{
+    (void)arg;
+    while (atomic_load(&is_running)) {
+        int *chunk = realloc(NULL, sizeof(int) * 1);
+        expect(chunk);
+        chunk = realloc(chunk, sizeof(int) * 2);
```
Comment (fjmolinas) [inline tests/malloc_thread_safety/main.c:66]: Can you address cppcheck comments? (or suppress when it makes sense)

File: tests/malloc_thread_safety/main.c [hunk 4]
```diff
@@ -78,41 +97,62 @@ int main(void)
...
+            failed = 1;
+            puts("Not all blocks were correctly freed!");
+            printf("mallinfo().uordblks before test: %u, after test: %u\n",
+                   (unsigned)pre.uordblks, (unsigned)post.uordblks);
+        }
 #endif
-    puts("TEST PASSED");
+    }
+
+    if (failed) {
```
Comment (fjmolinas) [inline tests/malloc_thread_safety/main.c:150]: If `ifndef NO_MALLINFO` this can never be false, does this global check make sense in that case? What is considered a passing test in that case?

**Review by fjmolinas — COMMENTED** [review]
Some comments

File: tests/malloc_thread_safety/Makefile (hunk 1 above)
Comment (maribu) [reply tests/malloc_thread_safety/Makefile:6]: Ah, yes. Since modeling both c libs as features, this workaround is no longer needed, right?

Comment (fjmolinas) [reply tests/malloc_thread_safety/Makefile:6]: And also this makes compilation currently fail in some case because both modules get selected.

**Comment by fjmolinas** [comment]
I had some test failures because the `picolibc` and `newlibc` module was selected twice, but this should be fixed by requiring the feature instead of adding the module, otherwise test passes with picolibc as well now https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/257/artifact/.

File: tests/malloc_thread_safety/main.c (hunk 4 above)
Comment (maribu) [reply tests/malloc_thread_safety/main.c:156]: The actual test is that this doesn't crash. E.g. without the mutex I get hard faults or `malloc()` never returning. The additional `failed` only checks that `malloc()` and friends don't leak memory.

Maybe I should drop this, since this would only detect bugs in the standard c lib rather than in RIOT. On the other hand, this check is pretty much for free and it might be useful to detect a broken toolchain. There also was some discussion on an alternative implementation in the context of the Google summer of code, which would profit from this test as well.

Comment (fjmolinas) [reply tests/malloc_thread_safety/main.c:156]: Makes sense, lets keep it mind adding a comment and suppressing the warning?

**Comment by maribu** [comment]
Time to squash?

**Review by fjmolinas — APPROVED** [review]
ACK, and GO

**Comment by maribu** [comment]
Thanks!
